cms/models/pagemodel.py

```diff
@@ -211,36 +211,51 @@ def move_page(self, target, position='first-child'):
+        assert target.publisher_is_draft
         # do not mark the page as dirty after page moves
         self._publisher_keep_state = True
 
         is_inherited_template = (
             self.template == constants.TEMPLATE_INHERITANCE_MAGIC)
+        target_public_page = None
+
+        if position in ('left', 'right') and not target.parent:
+            # make sure move_page does not break the tree
+            # when moving to a top level position.
+
+            if position == 'right' and target.publisher_public_id:
+                # The correct path order for pages at the root level
+                # is that any left sibling page has a path
+                # lower than the path of the current page.
+                # This rule applies to both draft and live pages
+                # so this condition will make sure to add the
+                # current page to the right of the target live page.
+                target_public_page = Page.objects.get(pk=target.publisher_public_id)
+                draft_root = target.get_root()
+                public_root = target_public_page.get_root()
+
+                # With this assert we can detect tree corruptions.
+                # It's important to raise this!
+                assert draft_root.get_next_sibling() == public_root
+
+            if is_inherited_template:
+                # The following code resolves the inherited template
+                # and sets it on the moved page.
+                # This is because the page is being moved to a root
+                # position and so can't inherit from anything.
+                self.template = self.get_template()
 
-        # make sure move_page does not break when using INHERIT template
-        # and moving to a top level position
-        if (position in ('left', 'right') and not target.parent and
-                is_inherited_template):
-            self.template = self.get_template()
-            if target.publisher_public_id and position == 'right':
-                public = target.publisher_public
-                target_root = target.get_root()
-                if target_root.get_next_sibling():
-                    if target_root.get_next_sibling() == public.get_root():
-                        target = public
-                    else:
-                        logger.warning('tree may need rebuilding: '
-                                       'run `manage.py cms fix-tree`')
-                else:
-                    # target's root is last 'root' node, switch to 'left'
-                    position = 'left'
         if position == 'first-child' or position == 'last-child':
             self.parent_id = target.pk
         else:
             self.parent_id = target.parent_id
 
         self.save()
-        moved_page = self.move(target, pos=position)
+
+        if target_public_page:
+            moved_page = self.move(target, pos=position)
+        else:
+            moved_page = self.move(target, pos=position)
 
```

Comment:
@thoreg Good catch, here is meant to be self.move(target_public_page, pos=position).